File: elixir/lib/2022/15_beacon_exclusion_zone/aoc202215.ex
defmodule AOC2022.Day15 do
  @moduledoc """
  Advent of Code 2022, day 15: Beacon Exclusion Zone.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input),
    do: puzzle_input |> String.split("\n", trim: true) |> Enum.map(&parse_line/1)

  @doc """
  Parse one line of input.

  ## Example:

      iex> parse_line("Sensor at x=2, y=18: closest beacon is at x=-2, y=15")
      {{2, 18}, {-2, 15}}
  """
  def parse_line(line) do
    %{"sx" => sx, "sy" => sy, "bx" => bx, "by" => by} =
      Regex.named_captures(
        ~r/Sensor at x=(?<sx>-?\d+), y=(?<sy>-?\d+): closest beacon is at x=(?<bx>-?\d+), y=(?<by>-?\d+)/,
        line
      )
      |> Enum.into(%{}, fn {key, pos} -> {key, String.to_integer(pos)} end)

    {{sx, sy}, {bx, by}}
  end

  @doc """
  Solve part 1.
  """
  def part1(sensors, row \\ 2_000_000) do
    coverage = sensors |> row_coverage(row) |> hd |> then(&Range.size(&1))

    beacons =
      sensors
      |> Enum.map(&elem(&1, 1))
      |> Enum.filter(&(elem(&1, 1) == row))
      |> Enum.into(%MapSet{}, &elem(&1, 0))
      |> MapSet.size()

    coverage - beacons
  end

  @doc """
  Solve part 2.
  """
  def part2(sensors, space \\ 4_000_000) do
    sensors |> find_hole(space) |> tuning_frequency()
  end

  @doc """
  Find sensor coverage for a given row.

  ## Example:

      6  ......211111111B..
      7  .....222111111111.
      8  ....2222211S111111
      9  ...222S2221111111.
      10 ....222221111111..
      11 .....2221111111...
      12 ......B..11111....

      iex> row_coverage([{{11, 8}, {15, 6}}, {{6, 9}, {6, 12}}], 12)
      [6..6, 9..13]
  """
  def row_coverage(sensors, row) do
    sensors
    |> Enum.map(fn {{sx, sy} = sensor, beacon} ->
      {sx, manhattan(sensor, beacon) - abs(row - sy)}
    end)
    |> Enum.reject(&(elem(&1, 1) < 0))
    |> Enum.map(fn {x, coverage} -> (x - coverage)..(x + coverage)//1 end)
    |> union()
  end

  @doc """
  Find a hole in the coverage.

  ## Example:

      TODO: Create a proper example
      5  ...........1......
      6  ......2...11B.....
      7  .....222.11111....
      8  ....2222211S111...
      9  ...222S2221111....
      10 ....22222.111.....
      11 .....222...1......
      12 ......B...........

      iex> find_hole([{{11, 8}, {12, 6}}, {{6, 9}, {6, 12}}], 20)
      {7, 6}
  """
  def find_hole(sensors, space) do
    distances = Enum.map(sensors, fn {sensor, beacon} -> {sensor, manhattan(sensor, beacon)} end)

    0..space
    |> Enum.find_value(fn row ->
      coverage =
        distances
        |> Enum.reject(fn {{_, sy}, distance} -> abs(row - sy) > distance end)
        |> Enum.map(fn {{sx, sy}, distance} ->
          max(0, sx - (distance - abs(row - sy)))..min(space, sx + (distance - abs(row - sy)))
        end)
        |> union()

      if length(coverage) > 1, do: {coverage |> hd |> then(&(&1.last + 1)), row}
    end)
  end

  # Slower because distances are calculated repeatedly
  # def find_hole(sensors, space) do
  #   0..space
  #   |> Enum.find_value(fn row ->
  #     coverage = row_coverage(sensors, row)
  #     if length(coverage) > 1, do: {coverage |> hd |> then(&(&1.last + 1)), row}
  #   end)
  # end

  @doc """
  Union of ranges.

  ## Examples:

      iex> union(2..4, 4..7)
      [2..7]
      iex> union(2..4, 5..7)
      [2..7]
      iex> union(-1..5, 8..12)
      [-1..5, 8..12]
      iex> union([-1..5, 8..12], 3..9)
      [-1..12]
      iex> union([-1..5, 8..12], 14..19)
      [-1..5, 8..12, 14..19]
      iex> union([])
      []
      iex> union([1..3])
      [1..3]
      iex> union([2..2, 9..9, 13..13, 14..14, 2..14, 2..2, -2..2, 16..24, 17..17])
      [-2..14, 16..24]
  """
  def union([]), do: []
  def union([range | ranges]), do: union(ranges, range)
  def union(one, two), do: _union(one, two) |> Enum.reverse()

  defp _union(ranges, range) when is_list(ranges) do
    [range | ranges]
    |> Enum.sort(fn one, two -> one.first <= two.first end)
    |> Enum.reduce([], fn
      range, [] -> [range]
      two, [one | ranges] -> _union(one, two) ++ ranges
    end)
  end

  defp _union(one, two) do
    if one.last + 1 < two.first or two.last + 1 < one.first,
      do: [two, one],
      else: [min(one.first, two.first)..max(one.last, two.last)]
  end

  @doc """
  Calculate the Manhattan distance between two points.

  ## Examples:

      iex> manhattan({3, 7}, {1, 11})
      6 = 2 + 4
      iex> manhattan({-2, 123}, {456, -12})
      593 = 458 + 135
  """
  def manhattan({x1, y1}, {x2, y2}), do: abs(x2 - x1) + abs(y2 - y1)

  @doc """
  Calculate tuning frequency.

  The tuning frequency can be found by multiplying the x coordinate by 4000000
  and then adding the y coordinate.

  ## Example:

      iex> tuning_frequency({14, 11})
      56_000_011
  """
  def tuning_frequency({x, y}), do: x * 4_000_000 + y

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2022/15_beacon_exclusion_zone/aoc202215.py
"""AoC 15, 2022: Beacon Exclusion Zone."""

# Standard library imports
import itertools
import pathlib
import re
import sys
from functools import reduce


def parse_data(puzzle_input):
    """Parse input.

    ## Example:

    >>> parse_data(
    ...     "Sensor at x=14, y=17: closest beacon is at x=10, y=16\\n"
    ...     "Sensor at x=8, y=7: closest beacon is at x=2, y=10"
    ... )
    ([(14, 17, 5), (8, 7, 9)], [(10, 16), (2, 10)])
    """
    sensors, beacons = zip(*[parse_sensor(line) for line in puzzle_input.split("\n")])
    size = 20 if max(sx for sx, _, _ in sensors) <= 30 else 4_000_000
    return list(sensors), list(beacons), size


def parse_sensor(line):
    """Parse information about one sensor.

    ## Example:

    >>> parse_sensor("Sensor at x=13, y=2: closest beacon is at x=15, y=3")
    ((13, 2, 3), (15, 3))
    """
    sx, sy, bx, by = map(int, re.findall(r"[xy]=(-?\d+)", line))
    d = abs(bx - sx) + abs(by - sy)
    return (sx, sy, d), (bx, by)


def part1(data):
    """Solve part 1."""
    sensors, beacons, size = data
    row = size // 2
    intervals = sorted(
        (x - (d - abs(row - y)), x + (d - abs(row - y)))
        for x, y, d in sensors
        if d >= abs(row - y)
    )
    coverage = sum(
        last - first + 1 for first, last in reduce(collapse_interval, intervals, [])
    )
    beacons_in_row = {bx for bx, by in beacons if by == row}
    return coverage - len(beacons_in_row)


def part2(data):
    """Solve part 2."""
    sensors, _, size = data

    # List diagonal lines one unit outside the borders of the diamonds
    nes = {xy for x, y, d in sensors for xy in [x + y + d + 1, x + y - d - 1]}
    ses = {xy for x, y, d in sensors for xy in [x - y + d + 1, x - y - d - 1]}

    # Loop over and check candidate intersection points
    for ne, se in itertools.product(nes, ses):
        x, y = (ne + se) // 2, (ne - se) // 2
        if (
            0 <= x <= size
            and 0 <= y <= size
            and all(abs(sx - x) + abs(sy - y) > d for sx, sy, d in sensors)
        ):
            return 4_000_000 * x + y


def collapse_interval(intervals, interval):
    """Collapse interval into intervals if it's overlapping.

    Assumes that intervals + [interval] is sorted.

    ## Example:

    >>> collapse_interval([(0, 2), (4, 7)], (5, 9))
    [(0, 2), (4, 9)]
    >>> collapse_interval([(1, 4)], (6, 11))
    [(1, 4), (6, 11)]
    >>> collapse_interval([(3, 9)], (4, 7))
    [(3, 9)]
    """
    if not intervals:
        return [interval]
    first_prev, last_prev = intervals[-1]
    first_new, last_new = interval
    if first_new <= last_prev + 1:
        return intervals[:-1] + [(first_prev, max(last_new, last_prev))]
    else:
        return intervals + [interval]


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
